### PR TITLE
141 update failure notification

### PR DESCRIPTION
**Pull request by WallysFerreira**
### Description

Change result message of email and slack notification in errors

- Relevant Issues : #141 
- Type of change :
  - [ ] New feature
  - [ ] Bug fix for existing feature
  - [x] Code quality improvement
  - [ ] Addition or Improvement of tests
  - [ ] Addition or Improvement of documentation

### CODE REVIEW

**Comment by WallysFerreira** [comment]
Should I add constants with new key strings?

**Comment by namitad** [comment]
@WallysFerreira yes - lets move them to constants

File: src/emailalert/email_function.py [hunk 1]
```diff
@@ -10,16 +10,18 @@ def lambda_handler(event, context):
     SUBJECT = "Failure alert for RDS Restore Pipeline"
     result = {}
     if 'status' in event:
-        result[constants.ERROR] = event['taskname'] + 'Error'
-        result[constants.CAUSE] = event['status']
+        result['failed step'] = event['taskname'] + 'Error'
+        result['cause of failure'] = event['status']
```
Comment (namitad) [inline src/emailalert/email_function.py:14]: lets use constants for these

File: src/emailalert/email_function.py [hunk 2]
```diff
@@ -10,16 +10,18 @@ def lambda_handler(event, context):
...
     elif 'Error' in event:
-        result[constants.ERROR] = event['Error']
-        result[constants.CAUSE] = event['Cause']
+        result['database id'] = event['DatabaseID']
```
Comment (namitad) [inline src/emailalert/email_function.py:16]: the event will contain `identifier` for database id and snapshot id is inferred from the database id. sharing a code reference https://github.com/intuit/Trapheus/blob/master/src/snapshot/snapshot_function.py#L11 

lets use the same here

File: src/emailalert/email_function.py [hunk 3]
```diff
@@ -10,16 +10,18 @@ def lambda_handler(event, context):
     SUBJECT = "Failure alert for RDS Restore Pipeline"
     result = {}
     if 'status' in event:
-        result[constants.ERROR] = event['taskname'] + 'Error'
-        result[constants.CAUSE] = event['status']
+        result['failed step'] = event['taskname'] + 'Error'
```
Comment (namitad) [inline src/emailalert/email_function.py:13]: this should be `result['failed step'] = event['taskname'] `

File: src/emailalert/email_function.py [hunk 4]
```diff
@@ -10,16 +10,18 @@ def lambda_handler(event, context):
...
+        result['cause of failure'] = event['status']
     elif 'Error' in event:
-        result[constants.ERROR] = event['Error']
-        result[constants.CAUSE] = event['Cause']
+        result['database id'] = event['DatabaseID']
+        result['snapshot id'] = event['SnapshotID']
+        result['failed step'] = event['Error']
```
Comment (namitad) [inline src/emailalert/email_function.py:18]: similar to above: `result['failed step'] = event['taskname']`

File: src/emailalert/email_function.py [hunk 5]
```diff
@@ -10,16 +10,18 @@ def lambda_handler(event, context):
...
+        result['cause of failure'] = event['Cause']
     BODY_HTML = """<html>
     <head></head>
     <body>
     <h1>Failure alert for RDS Restore Pipeline</h1>
-    <p>""" + result['Error'] + """ : """ + result['Cause'] + """</p>
+    <p>""" + result['failed step'] + """ : """ + result['cause of failure'] + """</p>
```
Comment (namitad) [inline src/emailalert/email_function.py:24]: this message should be updated to contain the message in the new format.

**Review by namitad — CHANGES_REQUESTED** [review]
have added some comments. please address

**Comment by namitad** [comment]
@WallysFerreira as part of the PR description, could you also add a screenshot of the new email and slack notification post your changes?

**Comment by WallysFerreira** [comment]
> @WallysFerreira as part of the PR description, could you also add a screenshot of the new email and slack notification post your changes?

Yes I can do that

File: src/emailalert/email_function.py [hunk 6]
```diff
@@ -10,16 +10,21 @@ def lambda_handler(event, context):
...
     result = {}
     if 'status' in event:
-        result[constants.ERROR] = event['taskname'] + 'Error'
+        result['failed step'] = event['taskname']
         result[constants.CAUSE] = event['status']
     elif 'Error' in event:
-        result[constants.ERROR] = event['Error']
+        result['database id'] = event['Identifier']
+        result['snapshot id'] = event['Identifier'] + constants.SNAPSHOT_POSTFIX
+        result['failed step'] = event['taskname']
```
Comment (namitad) [inline src/emailalert/email_function.py:18]: can we move all these keys to constants?

File: src/emailalert/email_function.py [hunk 7]
```diff
@@ -10,16 +10,21 @@ def lambda_handler(event, context):
     SUBJECT = "Failure alert for RDS Restore Pipeline"
     result = {}
     if 'status' in event:
-        result[constants.ERROR] = event['taskname'] + 'Error'
+        result['failed step'] = event['taskname']
         result[constants.CAUSE] = event['status']
     elif 'Error' in event:
-        result[constants.ERROR] = event['Error']
+        result['database id'] = event['Identifier']
```
Comment (namitad) [inline src/emailalert/email_function.py:16]: can we move the initialisation of `result['database id']` and `result['snapshot id']` outside of the if else loop as these will be common

File: src/slackNotification/slack_notification.py [hunk 8]
```diff
@@ -14,8 +14,10 @@ def lambda_handler(event, context):
         message[constants.ERROR] = event['taskname'] + 'Error'
         message[constants.CAUSE] = event['status']
     elif 'Error' in event:
-        message[constants.ERROR] = event['Error']
-        message[constants.CAUSE] = event['Cause']
+        message['database id'] = event['databaseid']
+        message['snapshot id'] = event['snapshotid']
```
Comment (namitad) [inline src/slackNotification/slack_notification.py:18]: 1. can we move the initialisation of `result['database id']` and `result['snapshot id']` outside of the if else loop as these will be common. 

2. here too - the snapshot id will be of the format `event['Identifier'] + constants.SNAPSHOT_POSTFIX`

**Comment by namitad** [comment]
@WallysFerreira could you check why the PR build on circleci is failing?

**Comment by WallysFerreira** [comment]
> @WallysFerreira could you check why the PR build on circleci is failing?

Yeah I will do that

**Comment by WallysFerreira** [comment]
circleci check is now passing, apparently there was an unit test failing and I think that was causing the check to fail too.

I'll add the screenshots you asked for soon

File: src/emailalert/email_function.py [hunk 9]
```diff
@@ -9,17 +9,24 @@ def lambda_handler(event, context):
...
     elif 'Error' in event:
-        result[constants.ERROR] = event['Error']
         result[constants.CAUSE] = event['Cause']
+
     BODY_HTML = """<html>
     <head></head>
     <body>
     <h1>Failure alert for RDS Restore Pipeline</h1>
-    <p>""" + result['Error'] + """ : """ + result['Cause'] + """</p>
+    <p>database id:""" + result['database id'] + """</p>
```
Comment (namitad) [inline src/emailalert/email_function.py:26]: can we use the constants here as well?

**Comment by namitad** [comment]
@WallysFerreira some minor comments. once you share the screenshots of the final email/slack, i will go ahead and approve the PR

**Comment by WallysFerreira** [comment]
I'm having some issues running the application on my computer to take the screenshots

**Comment by namitad** [comment]
@WallysFerreira could you elaborate - what issues are you facing?

**Comment by WallysFerreira** [comment]
Maybe I jumped some step on the tutorial, but when it tries to deploy to AWS it returns

Error: Template file not found at /home/wf/Trapheus/deploy.yaml

**Comment by namitad** [comment]
@WallysFerreira do reverify your setup and a give it a try again. If it doesnt work, let me know - i can try running your changes locally to verify the output.

**Comment by WallysFerreira** [comment]
I've tried running on a codespace and it's still returning an error, even though there is a template.yaml and deploy.yaml on the directory.
If you could run the changes locally I'd really appreciate it 

FileNotFoundError: [Errno 2] No such file or directory: 'sam package --template-file template.yaml --output-template-file deploy.yaml --s3-bucket teste-741284363623 --region us-east-2'

**Comment by namitad** [comment]
Ok.. give me a day or two. Will try this locally and get back to you.

**Comment by namitad** [comment]
@WallysFerreira i ran your changes locally. but i get the following error: 
```
{
  "errorMessage": "'Identifier'",
  "errorType": "KeyError",
  "requestId": "9c87e534-7933-4044-9295-d8a6834c23c8",
  "stackTrace": [
    "  File \"/var/task/email_function.py\", line 18, in lambda_handler\n    result[constants.DB_ID] = event['Identifier']\n"
  ]
}
```
Screenshot 2023-10-08 at 7 46 49 PM

can we check why `identifier` is not part of the input and if any changes need to be made in the template.yaml to cascade this input?

**Comment by WallysFerreira** [comment]
Yeah sure I'll look into it and use the new install.py. Thanks

**Comment by namitad** [comment]
@WallysFerreira i have merged the fix for install.py. can you pull the latest on your branch and try - no need to use a separate version of the file

**Comment by WallysFerreira** [comment]
I think the error was caused by the capital 'i' in identifier, but can't actually check because I'm getting another error when running install.py

Error: Unable to upload artifact src/common/common.zip referenced by ContentUri parameter of CommonLambdaLayer resource.
Unable to locate credentials

**Comment by namitad** [comment]
@WallysFerreira looks like your temporary aws credentials may have expired. could you refresh and try again?

**Comment by WallysFerreira** [comment]
I tried with new aws credentials and still got the same error :(
Could you please try running it locally?

**Comment by stationeros** [comment]
@namitad / @WallysFerreira there is a couple of issues in the way this is implemented. The way we are getting the identifier is a bit off. The event that gets generated is  this

```
{'Error': 'SnapshotCreationException', 'Cause': '{"errorMessage": "Identifier:database-3 \\nAn error occurred (DBClusterNotFoundFault) when calling the CreateDBClusterSnapshot operation: DBCluster not found: database-3", "errorType": "SnapshotCreationException", "requestId": "6568ebed-be45-4962-bdc3-35bf080c4322", "stackTrace": ["  File \\"/var/task/cluster_snapshot_function.py\\", line 25, in lambda_create_cluster_snapshot\\n    util.eval_snapshot_exception(error, cluster_id, rds)\\n", "  File \\"/opt/python/utility.py\\", line 72, in eval_snapshot_exception\\n    raise custom_exceptions.SnapshotCreationException(error_message)\\n"]}'} | event is {'Error': 'SnapshotCreationException', 'Cause': '{"errorMessage": "Identifier:database-3 \\nAn error occurred (DBClusterNotFoundFault) when calling the CreateDBClusterSnapshot operation: DBCluster not found: database-3", "errorType": "SnapshotCreationException", "requestId": "6568ebed-be45-4962-bdc3-35bf080c4322", "stackTrace": [" File \\"/var/task/cluster_snapshot_function.py\\", line 25, in lambda_create_cluster_snapshot\\n util.eval_snapshot_exception(error, cluster_id, rds)\\n", " File \\"/opt/python/utility.py\\", line 72, in eval_snapshot_exception\\n raise custom_exceptions.SnapshotCreationException(error_message)\\n"]}'}
```

As you can see there is no direct way of getting the Identifier as a key. We have a utilty function defined that needs to be utilised https://github.com/intuit/Trapheus/blob/master/src/common/python/utility.py#L33 to get the identifier from the event.

Secondly, the way ```TASK_NAME``` is getting extracted also is a bit odd. Its not present directly in the event but its in the context object of the lambda runtime. You can simply do context.function_name to get that.

With the above changes, you guys should be able to get that out. Let me know if you need help 

Screenshot 2023-10-15 at 3 32 47 PM

**Comment by namitad** [comment]
@WallysFerreira it would still run into a similar issue (as mentioned by @stationeros above) with the "identifier" change. we need to explicitly pass the identifier name to the state where alerts are being sent. Can we look at using a pass state https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-pass-state.html to add this?

This can also be used in the future if we need to add additional values - such as execution id.

**Comment by namitad** [comment]
@stationeros yes - taskname is being cascaded because in the context object - it would either be the email alert/slack notification lambda name. even if set from each step, it would expose the lambda function name to the user instead of the flow. we can revisit if this change needs to be done and maybe address that as a separate issue.

**Comment by stationeros** [comment]
I am fine with whatever you guys choose to decide on this. Also I feel along with the documentation , it would be helpful to point to the code as to how we are passing that state today via template.yaml for better understanding.

**Comment by WallysFerreira** [comment]
@stationeros that's super helpful, thanks. I'll make those changes asap.

@namitad I'll try to look into using a pass state afterwards, but it would still work without it if I make the changes @stationeros said, right? Just so I can do one thing at a time

**Comment by stationeros** [comment]
@WallysFerreira Thats right? You can possibly fix the instance identifier and the task name chaining can be done later. I can break this into subtasks for this to be managed better

**Comment by stationeros** [comment]
@WallysFerreira  Can you make the code change for instance identifier , you just need to call this inside the lambda , we can then merge this commit

**Comment by stationeros** [comment]
@WallysFerreira  Closing this PR due to inactivity, feel free to re-open and submit if you are still working on this.

**Pull request closed without merging**